compliance/collection/nodeinventorizer/cachedInventorizer.go

```diff
@@ -0,0 +1,163 @@
+	currentBackoff, err := getCurrentBackoff(opts)
+	if err != nil {
+		return nil, err
+	}
+	if *currentBackoff > initialBackoff {
+		log.Warnf("Found existing backoff - last scan may have failed. Waiting %v seconds before running next inventory", currentBackoff)
+		opts.BackoffWaitCallback(*currentBackoff)
+	}
+	nextBackoff := calcNextBackoff(*currentBackoff)
+	bErr := writeBackoff(nextBackoff, opts.BackoffFilePath)
+	if bErr != nil {
+		log.Warnf("Error while persisting inventory backoff interval: %v", bErr)
+	}
+
+	// check whether a cached inventory exists that is recent enough to use
+	cachedInventory := loadCachedInventory(opts)
+	if cachedInventory != nil && isInventoryInCacheDuration(cachedInventory) {
+		log.Debugf("Using cached scan created at %v", cachedInventory.GetScanTime())
+		return createAndObserveMessage(cachedInventory), nil
+	}
+
+	// if no inventory exists, or it is too old, collect a fresh one and save it to the cache
+	newInventory, err := opts.Scanner.Scan(opts.NodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	err = persistInventoryToCache(newInventory, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	removeBackoff(opts.BackoffFilePath) // Remove backoff file as late as possible
+	return createAndObserveMessage(newInventory), nil
+}
+
+func getCurrentBackoff(opts *InventoryScanOpts) (*time.Duration, error) {
+	backoffInterval := env.NodeInventoryInitialBackoff.DurationSetting()
+
+	backoffFileContents, err := os.ReadFile(opts.BackoffFilePath)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No backoff found, continuing without pause")
+		} else {
+			return nil, err
+		}
+	} else {
+		// We have an existing backoff counter
+		backoffInterval, err = time.ParseDuration(string(backoffFileContents))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &backoffInterval, nil
+}
+
+func writeBackoff(backoff time.Duration, path string) error {
+	err := os.WriteFile(path, []byte(backoff.String()), 0600)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func calcNextBackoff(currentBackoff time.Duration) time.Duration {
+	maxBackoff := env.NodeInventoryMaxBackoff.DurationSetting()
+	nextBackoffInterval := currentBackoff + env.NodeInventoryBackoffIncrement.DurationSetting()
+	if nextBackoffInterval > maxBackoff {
+		log.Debugf("Backoff interval hit upper boundary. Cutting from %v to %v", nextBackoffInterval, maxBackoff)
+		nextBackoffInterval = maxBackoff
+	}
+	return nextBackoffInterval
+}
+
+func removeBackoff(backoffFilePath string) {
+	if err := os.Remove(backoffFilePath); err != nil {
+		log.Warnf("Could not remove scan backoff state file: %v", err)
+	}
+}
+
+func loadCachedInventory(opts *InventoryScanOpts) *storage.NodeInventory {
+	var cachedInv *storage.NodeInventory
+
+	cacheContents, err := os.ReadFile(opts.InventoryCachePath)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No cache file found, running new inventory")
```

Comment:
```suggestion
			log.Debug("No cache file found, running new scan")
```

Reply:
As suggested: Could we address this in ROX-14358?